File: app/src/main/java/com/savvasdalkitsis/gameframe/saves/usecase/SavedDrawingUseCase.java
package com.savvasdalkitsis.gameframe.saves.usecase;

import android.support.annotation.NonNull;
import android.support.v4.util.Pair;

import com.savvasdalkitsis.gameframe.GameFrameApplication;
import com.savvasdalkitsis.gameframe.bmp.usecase.BmpUseCase;
import com.savvasdalkitsis.gameframe.grid.model.Grid;
import com.savvasdalkitsis.gameframe.saves.model.SavedDrawingAlreadyExistsException;

import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

import rx.Observable;

public class SavedDrawingUseCase {

    private final BmpUseCase bmpUseCase;
    private final GameFrameApplication application;

    public SavedDrawingUseCase(BmpUseCase bmpUseCase, GameFrameApplication application) {
        this.bmpUseCase = bmpUseCase;
        this.application = application;
    }

    public Observable<File> saveDrawing(String name, Grid colorGrid) {
        return file(name)
                .flatMap(dir -> {
                    if (dir.exists()) {
                        return Observable.error(new SavedDrawingAlreadyExistsException("The directory '" + name + "' already exists"));
                    }
                    if (!dir.mkdirs()) {
                        return Observable.error(new IOException("Could not create directory " + dir.getAbsolutePath()));
                    }
                    return Observable.just(new File(dir, "0.bmp"));
                })
                .zipWith(bmpUseCase.rasterizeToBmp(colorGrid), Pair::create)
                .flatMap(save -> {
                    try {
                        FileOutputStream stream = new FileOutputStream(save.first);
                        stream.write(save.second);
                        stream.close();
                        return Observable.just(save.first);
                    } catch (IOException e) {
                        return Observable.error(e);
                    }
                });
    }

    public Observable<Void> deleteDrawing(String name) {
        return file(name)
                .flatMap(dir -> {
                    if (!dir.exists()) {
                        return Observable.just(null);
                    }
                    try {
                        FileUtils.deleteDirectory(dir);
                        return Observable.just(null);
                    } catch (IOException e) {
                        return Observable.error(e);
                    }
                });
    }

    @NonNull
    private Observable<File> file(String name) {
        return Observable.just(new File(application.getExternalFilesDir(null), name));
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/saves/usecase/SavedDrawingUseCase.kt
package com.savvasdalkitsis.gameframe.saves.usecase

import com.savvasdalkitsis.gameframe.GameFrameApplication
import com.savvasdalkitsis.gameframe.bmp.usecase.BmpUseCase
import com.savvasdalkitsis.gameframe.grid.model.Grid
import com.savvasdalkitsis.gameframe.saves.model.SavedDrawingAlreadyExistsException
import org.apache.commons.io.FileUtils
import rx.Observable
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class SavedDrawingUseCase(private val bmpUseCase: BmpUseCase, private val application: GameFrameApplication) {

    fun saveDrawing(name: String, colorGrid: Grid): Observable<File> {
        return file(name)
                .flatMap { dir ->
                    when {
                        dir.exists() -> Observable.error<File>(SavedDrawingAlreadyExistsException("The directory '$name' already exists"))
                        !dir.mkdirs() -> Observable.error<File>(IOException("Could not create directory " + dir.absolutePath))
                        else -> Observable.just(File(dir, "0.bmp"))
                    }
                }
                .zipWith<ByteArray, Pair<File, ByteArray>>(bmpUseCase.rasterizeToBmp(colorGrid), { a, b -> Pair(a, b) })
                .flatMap { (file, bmp) ->
                    try {
                        val stream = FileOutputStream(file)
                        stream.write(bmp)
                        stream.close()
                        Observable.just(file)
                    } catch (e: IOException ) {
                        Observable.error<File>(e)
                    }
                }
    }

    fun deleteDrawing(name: String): Observable<Void> {
        return file(name)
                .flatMap { dir ->
                    if (!dir.exists()) {
                        Observable.just<Void>(null)
                    } else try {
                        FileUtils.deleteDirectory(dir)
                        Observable.just<Void>(null)
                    } catch (e: IOException) {
                        Observable.error<Void>(e)
                    }
                }
    }

    private fun file(name: String): Observable<File> {
        return Observable.just(File(application.getExternalFilesDir(null), name))
    }
}
